### ISSUE TITLE
IllegalArgumentException in RangeJoin

### ISSUE THREAD

**Issue by alifiyaali** [DataSystemsLab/GeoSpark#193]
While doing a range join, I am getting the following exception and not able to figure out the reason for same : 

> java.lang.IllegalArgumentException: Number of partitions must be >= 0
	at org.datasyslab.geospark.spatialRDD.SpatialRDD.spatialPartitioning(SpatialRDD.java:164)
	at org.apache.spark.sql.geosparksql.strategy.join.TraitJoinQueryExec$class.doSpatialPartitioning(TraitJoinQueryExec.scala:198)
	at org.apache.spark.sql.geosparksql.strategy.join.RangeJoinExec.doSpatialPartitioning(RangeJoinExec.scala:44)
	at org.apache.spark.sql.geosparksql.strategy.join.TraitJoinQueryExec$class.doExecute(TraitJoinQueryExec.scala:127)
	at org.apache.spark.sql.geosparksql.strategy.join.RangeJoinExec.doExecute(RangeJoinExec.scala:44)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:114)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:114)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:135)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:132)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:113)
	at org.apache.spark.sql.execution.SparkPlan.getByteArrayRdd(SparkPlan.scala:225)
	at org.apache.spark.sql.execution.SparkPlan.executeTake(SparkPlan.scala:308)
	at org.apache.spark.sql.execution.CollectLimitExec.executeCollect(limit.scala:38)
	at org.apache.spark.sql.Dataset$$anonfun$org$apache$spark$sql$Dataset$$execute$1$1.apply(Dataset.scala:2371)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:57)
	at org.apache.spark.sql.Dataset.withNewExecutionId(Dataset.scala:2765)
	at org.apache.spark.sql.Dataset.org$apache$spark$sql$Dataset$$execute$1(Dataset.scala:2370)
	at org.apache.spark.sql.Dataset.org$apache$spark$sql$Dataset$$collect(Dataset.scala:2377)
	at org.apache.spark.sql.Dataset$$anonfun$head$1.apply(Dataset.scala:2113)
	at org.apache.spark.sql.Dataset$$anonfun$head$1.apply(Dataset.scala:2112)
	at org.apache.spark.sql.Dataset.withTypedCallback(Dataset.scala:2795)
	at org.apache.spark.sql.Dataset.head(Dataset.scala:2112)
	at org.apache.spark.sql.Dataset.take(Dataset.scala:2327)
	at org.apache.spark.sql.Dataset.showString(Dataset.scala:248)
	at org.apache.spark.sql.Dataset.show(Dataset.scala:638)
	at com.inn.sparkrunner.processors.transformation.ExecuteQuery.execute(ExecuteQuery.java:117)

Here is the **physical plan** for the query :
> |== Physical Plan ==
*Project [latitude#5283, longitude#5284, type#5285, neighbour#2644, childLat#2, childLon#3, parentLon#0, parentLat#1]
+- RangeJoin st_geomfromwkt(regexp_replace(regexp_replace(polygon#2645, \s*\[\s*([-?0-9.]*)\s*\,\s*([-?0-9.]*)\s*\]\s*, $1 $2), \[(.*)\], POLYGON(($1)))), st_point(cast(longitude#5284 as decimal(16,10)), cast(latitude#5283 as decimal(16,10))), false
   :- Generate posexplode(UDF(childLat#2, childLon#3, 1000.0, 2.0, 0.0, 60.0, 6)), true, false, [neighbour#2644, polygon#2645]
   :  +- *FileScan parquet [parentLon#0,parentLat#1,childLat#2,childLon#3] Batched: true, Format: Parquet, Location: InMemoryFileIndex[file:/home/ist/Pictures/Product_ODSC/BuildingParquet], PartitionFilters: [], PushedFilters: [], ReadSchema: struct<parentLon:double,parentLat:double,childLat:double,childLon:double>
   +- Scan ExistingRDD[latitude#5283,longitude#5284,type#5285]|

**Comment by jiayuasu** [comment]
@alifiyaali Can you show me your SparkSQL code?

**Comment by alifiyaali** [comment]
@jiayuasu The code base is quite large. Can you point out some possible reasons which cause this exception. So that I can try to debug or provide you with some additional information?

**Comment by alifiyaali** [comment]
@jiayuasu 
The query is -
> select * from zone z, building b where ST_Intersects(z.polygon, b.polygon);

The following are some test cases :

1. When zone partition = 200 and building partition = 1 : Exception occurs

2. When zone partition = 200 and building partition = 200 : Exception occurs (Since there are empty partitions)

3. When zone partition = 1 and building partition = 1 : Query works

Can you suggest me a workaround rather than coalesce or repartition?

**Comment by tociek** [comment]
If you swap b and z parameters in the ST function call, do you still get the exception?

**Comment by alifiyaali** [comment]
@tociek I get a different exception now : 

> Caused by: java.lang.IllegalArgumentException: Points of LinearRing do not form a closed linestring
	at com.vividsolutions.jts.geom.LinearRing.validateConstruction(LinearRing.java:111)
	at com.vividsolutions.jts.geom.LinearRing.<init>(LinearRing.java:106)
	at com.vividsolutions.jts.geom.GeometryFactory.createLinearRing(GeometryFactory.java:341)
	at com.vividsolutions.jts.geom.GeometryFactory.createLinearRing(GeometryFactory.java:328)
	at com.vividsolutions.jts.io.WKTReader.readLinearRingText(WKTReader.java:549)
	at com.vividsolutions.jts.io.WKTReader.readPolygonText(WKTReader.java:635)
	at com.vividsolutions.jts.io.WKTReader.readGeometryTaggedText(WKTReader.java:481)
	at com.vividsolutions.jts.io.WKTReader.read(WKTReader.java:205)
	at com.vividsolutions.jts.io.WKTReader.read(WKTReader.java:174)
	at org.datasyslab.geospark.formatMapper.FormatMapper.readWkt(FormatMapper.java:111)
	at org.datasyslab.geospark.formatMapper.FormatMapper.readGeometry(FormatMapper.java:140)
	at org.apache.spark.sql.geosparksql.expressions.ST_GeomFromWKT.eval(Constructors.scala:156)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.apply_1$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.apply(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.apply(Unknown Source)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
	at scala.collection.Iterator$class.foreach(Iterator.scala:893)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
	at scala.collection.TraversableOnce$class.foldLeft(TraversableOnce.scala:157)
	at scala.collection.AbstractIterator.foldLeft(Iterator.scala:1336)
	at scala.collection.TraversableOnce$class.aggregate(TraversableOnce.scala:214)
	at scala.collection.AbstractIterator.aggregate(Iterator.scala:1336)
	at org.apache.spark.rdd.RDD$$anonfun$aggregate$1$$anonfun$22.apply(RDD.scala:1112)
	at org.apache.spark.rdd.RDD$$anonfun$aggregate$1$$anonfun$22.apply(RDD.scala:1112)
	at org.apache.spark.SparkContext$$anonfun$33.apply(SparkContext.scala:1980)
	at org.apache.spark.SparkContext$$anonfun$33.apply(SparkContext.scala:1980)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
	at org.apache.spark.scheduler.Task.run(Task.scala:99)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:282)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

b and z both are Polygon Geometry constructed using ST_GeomFromWKT , still I don't understand why is LinearRing being created?

**Comment by alifiyaali** [comment]
No, I don't get the exception after swapping. But the performance has become extremely poor as data in **b** is very large as compared to **z**.

**Comment by jiayuasu** [comment]
@alifiyaali This is a very interesting phenomenon. I roughly know what's going on but all my current test data and test cases cannot re-produce it.

To bypass this error, you should force the spatial join query to use a fallback PartitionNum.
```
sparkSession = SparkSession.builder().
config("spark.serializer",classOf[KryoSerializer].getName).
config("spark.kryo.registrator", classOf[GeoSparkKryoRegistrator].getName).
config("geospark.join.numpartition",1000)
```

**geospark.join.numpartition** will repartition both datasets to 1000 partitions and join them. You can specify any integer you want.

**Comment by jiayuasu** [comment]
@alifiyaali 

It seems your code is wrong. 

```
select name,max(ST_GeomFromWKT(polygon)) as polygon 
from polygonCSV 
group by name
```

What is the meaning of max(ST_Geom)?? The max of the geometries? It doesn't make sense.

**Comment by jiayuasu** [comment]
@alifiyaali 

If your code is like
```
SELECT name, ST_GeomFromWKT(polygon) as polygon
FROM polygoncsv
```
Then register table polygoncsv

```
SELECT ST_Point(cast(1.0 as Decimal(16,10)),cast(2.0 as Decimal(16,10))) as point
```
Then register table points

```
SELECT *
FROM points p1, polygonCSV p2
WHERE ST_Within(p1.point,p2.polygon)
```

It should be fine. We have tested this many times.

**Comment by alifiyaali** [comment]
@jiayuasu The exception is generated when one of the dataset has been shuffled. To represent a simpler code I did that group by intentionally.

**Comment by alifiyaali** [comment]
Try using first() instead of max().

**Comment by jiayuasu** [comment]
@alifiyaali The code still works fine on my side... I don't know what's going on. Did you try the same code?

**Comment by tociek** [comment]
What version of spark are you using? I was getting this error with 2.2 but not with 2.1...

**Comment by alifiyaali** [comment]
Spark version : 2.1.0
Scala version : 2.11
GeoSpark version : 1.0.1

**Comment by alifiyaali** [comment]
@jiayuasu Yes, I have tried it and got the error.

**Comment by netchkin** [comment]
Same exception happened to me when I tried to join a Dataframe of 1 record with another. When I forced `config("geospark.join.numpartition",200)`, I got `java.lang.IllegalArgumentException: [GeoSpark] Number of partitions 200 cannot be larger than half of total records num 1`. I just wonder, can this be related? Perhaps at some point when spatial join does not get enough data to fill all the partitions?

**Comment by jonrein** [comment]
I got the same exception when trying to join two relatively small sample DataFrames.  Setting `geospark.join.numpartition` to a sufficiently low value solved the problem, but it would be nice to not have to do that.

**Comment by jiayuasu** [comment]
@jonrein The exception @netchkin got 
```
java.lang.IllegalArgumentException: [GeoSpark] Number of partitions 200 cannot be larger than half of total records num 1
```
is different from the original exception in this post
```
java.lang.IllegalArgumentException: Number of partitions must be >= 0
```

Which one did you get?

I don't suggest you repartition DataFrame. If you don't repartition DataFrame, you will not need to use `geospark.join.numpartition` and GeoSparkSQL will figure out the num partition by itself.

**Comment by jonrein** [comment]
@jiayuasu,
I got the original exception.  If I'm reading @netchkin's comment correctly, he also got the original exception before modifying `geospark.join.numpartition` to investigate.

I'm not repartitioning the DataFrame, as far as I can tell.  Here's an example with some dummy data that generates the `java.lang.IllegalArgumentException: Number of partitions must be >= 0` exception, but runs fine with `geospark.join.numpartition` set to 1:

`    

    var sparkSession = SparkSession.builder().
      config("spark.serializer",classOf[KryoSerializer].getName).
      config("spark.kryo.registrator", classOf[GeoSparkKryoRegistrator].getName).
      master("local[*]").appName("geoSparkTest").getOrCreate()
    GeoSparkSQLRegistrator.registerAll(sparkSession.sqlContext)

    val ll_df = sparkSession.createDataFrame(
      sparkSession.sparkContext.parallelize(
        Seq(Row(1, "40.0", "-120.0"), Row(2, "30.0", "-110.0"), Row(3, "20.0", "-100.0"))),
      StructType(
        List(StructField("id", IntegerType, true), StructField("lat", StringType, true), StructField("lon", StringType, true))
      ))
    ll_df.createOrReplaceTempView("ll_df")

    val pointDF = sparkSession.sql("select id, ST_Point(cast(ll_df.lat as Decimal(24,20)), cast(ll_df.lon as Decimal(24,20))) AS latlon_point FROM ll_df")
    pointDF.createOrReplaceTempView("point_df")

    val boundMaxMinsDF = sparkSession.createDataFrame(
      sparkSession.sparkContext.parallelize(
        Seq(Row("A", 25.0, -115.0, 35.0, -105.0), Row("B", 25.0, -135.0, 35.0, -125.0))),
      StructType(
        List(StructField("id", StringType, true), StructField("latmin", DoubleType, true),
          StructField("lonmin", DoubleType, true), StructField("latmax", DoubleType, true),
          StructField("lonmax", DoubleType, true))
      ))
    boundMaxMinsDF.createOrReplaceTempView("bm_df")

    val polygonEnvelopeDF = sparkSession.sql("select id, ST_PolygonFromEnvelope(" +
      "cast(bm_df.latmin as Decimal(24,20)), cast(bm_df.lonmin as Decimal(24,20)), " +
      "cast(bm_df.latmax as Decimal(24,20)), cast(bm_df.lonmax as Decimal(24,20)), bm_df.id) AS polygon FROM bm_df")
    polygonEnvelopeDF.createOrReplaceTempView("poly_env_df")

    val withinEnvelopeDF = sparkSession.sql("select * FROM point_df, poly_env_df WHERE ST_Within(point_df.latlon_point, poly_env_df.polygon)")
    withinEnvelopeDF.explain()
    withinEnvelopeDF.show()`

**Comment by jiayuasu** [comment]
@alifiyaali @tociek @netchkin @jonrein 

Thanks for the help from all of you!

I was able to re-produce this bug and found the reason of this bug. The partition number detection logic will fail at super small test data. See PR https://github.com/DataSystemsLab/GeoSpark/pull/207.

This patch is now available at GeoSpark 1.1.1-SNAPSHOT for spark 2.1, 2.2 and 2.3. Please read [1] to learn how to use SNAPSHOT versions.

[1] http://datasystemslab.github.io/GeoSpark/download/GeoSpark-All-Modules-Maven-Central-Coordinates/

**Comment by alifiyaali** [comment]
@jiayuasu I am unable to download the snapshot version. Please suggest some resolution. I am using maven version 3.5

**POM.XML**

><project\>
>.
>.
><dependency\>
>      <groupId\>org.datasyslab</groupId\>
>     <artifactId\>geospark-sql_2.1</artifactId\>
>     <version\>1.1.1-SNAPSHOT</version\>
></dependency\>
>.
>.
><profiles\>
>    <profile\>
>        <id\>allow-snapshots</id\>
>        <activation\><activeByDefault\>true</activeByDefault\></activation\>
>        <repositories\>
>            <repository\>
>                <id\>snapshots-repo</id\>
>                <url\>https://oss.sonatype.org/content/repositories/snapshots</url\>
>                <releases\><enabled\>false</enabled\></releases\>
>                <snapshots\><enabled\>true</enabled\></snapshots\>
>            </repository\>
>        </repositories\>
>    </profile\>
></profiles\>
>.
>.
></project\>

**Comment by jiayuasu** [comment]
@alifiyaali Not sure what's going on. The jar is here for sure: https://oss.sonatype.org/content/repositories/snapshots/org/datasyslab/geospark-sql_2.1/1.1.1-SNAPSHOT/

The coordinate is correct.

In the worst case, use JitPack instead. It will compile GeoSpark-for-Spark-2.1 branch on-the-fly and put all modules in a single jar.

https://jitpack.io/#DataSystemsLab/GeoSpark/GeoSpark-for-Spark-2.1-SNAPSHOT

**Comment by alifiyaali** [comment]
@jiayuasu Please check -

> [ERROR] An error occurred during dependency resolution of the following artifact:
    org.datasyslab:geospark-sql_2.1:1.1.1-SNAPSHOT
Caused by: Unable to get dependency information: Unable to read the metadata file for artifact 'org.datasyslab:geospark-sql_2.1:jar': Cannot find parent: org.datasyslab:geospark-parent for project: null:geospark-sql_2.1:jar:1.1.1-SNAPSHOT for project null:geospark-sql_2.1:jar:1.1.1-SNAPSHOT
  org.datasyslab:geospark-sql_2.1:jar:1.1.1-SNAPSHOT

>from the specified remote repositories:
  central (http://repo1.maven.org/maven2),
  snapshots-repo (https://oss.sonatype.org/content/repositories/snapshots)

**Comment by jiayuasu** [comment]
@alifiyaali Thanks for the information. There was a version mismatch between the parent project and submodule. I have fixed it and now you can download the SNAPSHOT version.

**Comment by alifiyaali** [comment]
@jiayuasu I successfully downloaded the SNAPSHOT and the issue is resolved. Thank You!

**Comment by td-truedetective** [comment]
@jiayuasu 

Is this issue resolved in geospark-sql_2.3:1.1.3 version as well?

Because I am getting this issue in geospark-sql_2.3:1.1.3

Here is the version detail - 

Spark version : **2.3.1.3.0.2.0-50** (hortonworks)
Scala version : **2.11**
GeoSpark version : **1.1.3**
**Java 8**

**Comment by jiayuasu** [comment]
@td-truedetective This bug has been fixed in 1.1.3. Are you sure that you got the exact error?
